### ISSUE TITLE
[ENHANCEMENT] Add LoadBalancer resource, data source to terraform-provider-harvester

### ISSUE THREAD

**Issue by connorkuehl** [harvester/harvester#4814]
**Is your enhancement related to a problem? Please describe.**

 It's not possible to describe LoadBalancers with OpenTofu/Terraform since the provider does not yet have support for it.

**Describe the solution you'd like**

Add the harvester_loadbalancer resource and data source to the provider

**Describe alternatives you've considered**

**Additional context**

**Comment by gregkonush** [comment]
declarative resource fro load balancer will be helpful for automation

**Comment by Vicente-Cheng** [comment]
Hi @m-ildefons, could you move this one to the correct status?
I thought we could use #8688 to validate all the terraform-provider enhancements

**Comment by Vicente-Cheng** [comment]
@m-ildefons, as we discussed offline
Please move this one to the Ready for Testing and update the test plan.

cc @innobead

**Comment by m-ildefons** [comment]
Waiting on https://github.com/harvester/terraform-provider-harvester/pull/127 before it can be moved to ready-for-testing.

**Comment by Vicente-Cheng** [comment]
Hi @m-ildefons, 
Please move to the correct status and update the test plan since https://github.com/harvester/terraform-provider-harvester/pull/127 was merged.

Thanks!
cc @innobead

**Comment by asc5543** [comment]
## Test result
Test FAIL, the terraform can't run.

## Environment
* Harvester ISO version: `v1.6.0-rc5`
* Terrafrom-provider-harvester: [f69ca13](https://github.com/harvester/terraform-provider-harvester/commit/f69ca13e2035f05a2a6f46cb936433a5d00d982e)
* Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): `1 node via ipxe-example`


## Test Scenario
### Preparation
1. Prepare the v1.6.0-rc5 harvester

  <img width="1913" height="982" alt="Image" src="https://github.com/user-attachments/assets/a848f635-39f5-4196-89c5-89bbf80d7922" />

2. Download the kubeconfig of the harvester, setup the version.tf for terraform

   ```
   terraform {
     required_providers {
       harvester = {
         source  = "harvester/harvester"
         version = "0.0.0-dev"
       }
     }
   }

   provider "harvester" {
     kubeconfig = "~/Documents/kubeconfig/local.yaml"
   }

   provider "kubernetes" {
     config_path    = "~/Documents/kubeconfig/local.yaml"
     config_context = "local"
   }
   ```


### Test Steps
1. <details>
   <summary> ✔ Prepare the test.tf for uploading image, create vm, and using loadbalancer resource (copy from examples)
   </summary>

   ```
   resource "harvester_image" "ubuntu20" {
     name      = "ubuntu20"
     namespace = "harvester-public"

     display_name = "ubuntu-20.04-server-cloudimg-amd64.img"
     source_type  = "download"
     url          = "http://cloud-images.ubuntu.com/releases/focal/release/ubuntu-20.04-server-cloudimg-amd64.img"
   }

   resource "harvester_virtualmachine" "ubuntu20" {
     name                 = "ubuntu20"
     namespace            = "default"
     restart_after_update = true

     description = "test ubuntu20 raw image"
     tags = {
       ssh-user = "ubuntu"
     }

     cpu    = 2
     memory = "2Gi"

     run_strategy    = "RerunOnFailure"
     hostname        = "ubuntu20"
     reserved_memory = "100Mi"
     machine_type    = "q35"

     network_interface {
       name           = "nic-1"
       wait_for_lease = true
     }

     disk {
       name       = "rootdisk"
       type       = "disk"
       size       = "10Gi"
       bus        = "virtio"
       boot_order = 1

       image       = harvester_image.ubuntu20.id
       auto_delete = true
     }

     disk {
       name        = "emptydisk"
       type        = "disk"
       size        = "20Gi"
       bus         = "virtio"
       auto_delete = true
     }

   }


   resource "harvester_loadbalancer" "service_loadbalancer" {
     name = "service-loadbalancer"

     depends_on = [
       harvester_virtualmachine.ubuntu20
     ]

     listener {
       port         = 443
       protocol     = "tcp"
       backend_port = 8080
     }

     listener {
       port         = 80
       protocol     = "tcp"
       backend_port = 8080
     }

     ipam   = "ippool"
     ippool = "service-ips"

     workload_type = "vm"

     backend_selector {
       key    = "app"
       values = ["test"]
     }

     backend_selector {
       key    = "component"
       values = ["frontend", "ui"]
     }

     healthcheck {
       port              = 443
       success_threshold = 1
       failure_threshold = 3
       period_seconds    = 10
       timeout_seconds   = 5
     }
   }
   ```

   </details>

1. <details open>
   <summary> ✗ Trigger terraform init and terraform plan, it shows error <b>Error: expected ipam to be one of [pool dhcp], got ippool</b>
   </summary>
  
   ```
   Error: expected ipam to be one of [pool dhcp], got ippool

     on test.tf line 74, in resource "harvester_loadbalancer" "service_loadbalancer":
     74:   ipam   = "ippool"
   ```

   It means the examples of [loadbalancer](https://github.com/harvester/terraform-provider-harvester/blob/master/examples/resources/harvester_loadbalancer/resources.tf) should be updated

   </details>

1. <details open>
   <summary> ✗ Change ipam to <b>pool</b>, then trigger <b>terraform init</b>, <b>terraform apply</b>, then the error shows: <b>Error: invalid value for IPAM: pool</b>
   </summary>

   ```
   Error: invalid value for IPAM: pool

     on test.tf line 55, in resource "harvester_loadbalancer" "service_loadbalancer":
     55: resource "harvester_loadbalancer" "service_loadbalancer" {
   ```

   The fail reason is that in our script, we only support "dhcp" and "cluster"
   ```
   func (c *Constructor) subresourceLoadBalancerIPAMParser(data interface{}) error {
   	ipam := data.(string)

   	if ipam != "dhcp" && ipam != "cluster" {
   		return fmt.Errorf("invalid value for IPAM: %v", ipam)
   	}

   	c.LoadBalancer.Spec.IPAM = loadbalancerv1.IPAM(ipam)
	return nil
   }
   ```

   </details>

1. <details open>
   <summary> ✗ Change ipam to <b>dhcp</b>, then trigger <b>terraform init</b>, <b>terraform apply</b>, then the Error shows: <b>admission webhook "validator.harvester-system.harvester-load-balancer-webhook" denied the request: Internal error occurred: create loadbalancer default/service-loadbalancer failed: listener has duplicate name </b>
   </summary>

   ```
   harvester_loadbalancer.service_loadbalancer: Creating...
   harvester_image.ubuntu20: Modifying... [id=harvester-public/ubuntu20]
   harvester_image.ubuntu20: Modifications complete after 0s [id=harvester-public/ubuntu20]

   Error: admission webhook "validator.harvester-system.harvester-load-balancer-webhook" denied the request: Internal error occurred: create loadbalancer default/service-loadbalancer failed: listener has duplicate name 

     on test.tf line 55, in resource "harvester_loadbalancer" "service_loadbalancer":
     55: resource "harvester_loadbalancer" "service_loadbalancer" {


   ```

   </details>

@m-ildefons can you help to check this? thank you.
It seems like the examples need to be updated

**Comment by m-ildefons** [comment]
Hi Justin,
thanks for checking this. Seems like there is still a bit of fixup needed.
I've made another PR with some small changes: https://github.com/harvester/terraform-provider-harvester/pull/129
It also includes improvements to the examples.

**Comment by m-ildefons** [comment]
Moving this back to ready-for-testing since the fixes are merged now.
@asc5543 could you please check again? There is also an extensive example in the PR https://github.com/harvester/terraform-provider-harvester/pull/129

**Comment by asc5543** [comment]
## Test result
Test PASS, the terraform-provider-harvester can run without issue

## Environment
* Harvester ISO version: `v1.6.0-rc5`
* Terrafrom-provider-harvester: [[f691469](https://github.com/harvester/terraform-provider-harvester/commit/f691469f89ce924cd66115d823931d999933b5e0)
* Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): `2 node via ipxe-example`


## Test Scenario
### Preparation
1. Prepare the v1.6.0-rc5 harvester

<img width="1918" height="1078" alt="Image" src="https://github.com/user-attachments/assets/3176a515-fb95-4613-895a-76cf9e850f44" />

2. Download the kubeconfig of the harvester, setup the version.tf for terraform

   ```
   terraform {
     required_providers {
       harvester = {
         source  = "harvester/harvester"
         version = "0.0.0-dev"
       }
     }
   }

   provider "harvester" {
     kubeconfig = "~/Documents/kubeconfig/local.yaml"
   }

   provider "kubernetes" {
     config_path    = "~/Documents/kubeconfig/local.yaml"
     config_context = "local"
   }
   ```


### Test Steps
1. <details>
   <summary> ✔ Prepare the test.tf for uploading image, create vm, and using loadbalancer resource (copy from examples)
   </summary>

   ```
   resource "harvester_image" "ubuntu20" {
     name      = "ubuntu20"
     namespace = "harvester-public"

     display_name = "ubuntu-20.04-server-cloudimg-amd64.img"
     source_type  = "download"
     url          = "http://cloud-images.ubuntu.com/releases/focal/release/ubuntu-20.04-server-cloudimg-amd64.img"
   }

   resource "harvester_virtualmachine" "ubuntu20" {
     name                 = "ubuntu20"
     namespace            = "default"
     restart_after_update = true

     description = "test ubuntu20 raw image"
     tags = {
       ssh-user = "ubuntu"
       app = "test"
       component = "frontend"
     }

     cpu    = 2
     memory = "2Gi"

     run_strategy    = "RerunOnFailure"
     hostname        = "ubuntu20"
     reserved_memory = "100Mi"
     machine_type    = "q35"

     network_interface {
       name           = "nic-1"
       wait_for_lease = true
     }

     disk {
       name       = "rootdisk"
       type       = "disk"
       size       = "10Gi"
       bus        = "virtio"
       boot_order = 1

       image       = harvester_image.ubuntu20.id
       auto_delete = true
     }

     disk {
       name        = "emptydisk"
       type        = "disk"
       size        = "20Gi"
       bus         = "virtio"
       auto_delete = true
     }

     cloudinit {
       user_data = <<EOT
   #cloud-config
   user: ubuntu
   password: foobar
   chpasswd:
     expire: false
   packages:
     - qemu-guest-agent
     - nginx
   runcmd:
     - - systemctl
       - enable
       - --now
       - qemu-guest-agent.service
     - - systemctl
       - enable
       - --now
       - nginx.service
   EOT
     }
   }

   resource "harvester_ippool" "service_ips" {
     name = "service-ips"

     range {
       start   = "192.168.0.70"
       end     = "192.168.0.80"
       subnet  = "192.168.0.1/24"
       gateway = "192.168.0.1"
     }

     selector {
       priority = 100
       network  = "default/vm-network"
       scope {
         project       = "services"
         namespace     = "default"
         guest_cluster = "prod-services"
       }
     }
   }

   resource "harvester_loadbalancer" "service_loadbalancer_pool" {
     name = "service-loadbalancer-pool"

     depends_on = [
       harvester_virtualmachine.ubuntu20
     ]

     listener {
       name         = "http"
       port         = 80
       protocol     = "tcp"
       backend_port = 80
     }

     ipam   = "pool"
     ippool = "service-ips"

     workload_type = "vm"

     backend_selector {
       key    = "harvesterhci.io/vmName"
       values = ["ubuntu20"]
     }

     healthcheck {
       port              = 80
       success_threshold = 1
       failure_threshold = 3
       period_seconds    = 10
       timeout_seconds   = 5
     }
   }

   resource "harvester_loadbalancer" "service_loadbalancer_dhcp" {
     name = "service-loadbalancer-dhcp"

     depends_on = [
       harvester_virtualmachine.ubuntu20
     ]

     listener {
       name         = "http"
       port         = 80
       protocol     = "tcp"
       backend_port = 80
     }

     ipam   = "dhcp"

     workload_type = "vm"

     backend_selector {
       key    = "harvesterhci.io/vmName"
       values = ["ubuntu20"]
     }

     healthcheck {
       port              = 80
       success_threshold = 1
       failure_threshold = 3
       period_seconds    = 10
       timeout_seconds   = 5
     }
   }
   ```

   </details>

1. <details>
   <summary> ✔ Trigger terraform init, the init success without issue
   </summary>
  
   <img width="936" height="521" alt="Image" src="https://github.com/user-attachments/assets/57215174-ddc3-49c9-9525-6f629e2e66c3" />

   </details>

1. <details>
   <summary> ✔ Trigger <b>terraform apply</b>, and there is no error message shows</summary>

   ```
   harvester_image.ubuntu20: Creating...
   harvester_ippool.service_ips: Creating...
   harvester_ippool.service_ips: Creation complete after 0s [id=service-ips]
   harvester_image.ubuntu20: Still creating... [10s elapsed]
   harvester_image.ubuntu20: Still creating... [20s elapsed]
   harvester_image.ubuntu20: Still creating... [30s elapsed]
   harvester_image.ubuntu20: Still creating... [40s elapsed]
   harvester_image.ubuntu20: Still creating... [50s elapsed]
   harvester_image.ubuntu20: Still creating... [1m0s elapsed]
   harvester_image.ubuntu20: Still creating... [1m10s elapsed]
   harvester_image.ubuntu20: Still creating... [1m20s elapsed]
   harvester_image.ubuntu20: Still creating... [1m30s elapsed]
   harvester_image.ubuntu20: Still creating... [1m40s elapsed]
   harvester_image.ubuntu20: Still creating... [1m50s elapsed]
   harvester_image.ubuntu20: Still creating... [2m0s elapsed]
   harvester_image.ubuntu20: Still creating... [2m10s elapsed]
   harvester_image.ubuntu20: Still creating... [2m20s elapsed]
   harvester_image.ubuntu20: Still creating... [2m30s elapsed]
   harvester_image.ubuntu20: Still creating... [2m40s elapsed]
   harvester_image.ubuntu20: Still creating... [2m50s elapsed]
   harvester_image.ubuntu20: Creation complete after 2m51s [id=harvester-public/ubuntu20]
   harvester_virtualmachine.ubuntu20: Creating...
   harvester_virtualmachine.ubuntu20: Still creating... [10s elapsed]
   harvester_virtualmachine.ubuntu20: Still creating... [20s elapsed]
   harvester_virtualmachine.ubuntu20: Still creating... [30s elapsed]
   harvester_virtualmachine.ubuntu20: Creation complete after 39s [id=default/ubuntu20]
   harvester_loadbalancer.service_loadbalancer_dhcp: Creating...
   harvester_loadbalancer.service_loadbalancer_pool: Creating...
   harvester_loadbalancer.service_loadbalancer_pool: Creation complete after 0s [id=default/service-loadbalancer-pool]
   harvester_loadbalancer.service_loadbalancer_dhcp: Creation complete after 0s [id=default/service-loadbalancer-dhcp]

   Apply complete! Resources: 5 added, 0 changed, 1 destroyed.
   ```

   </details>

1. <details>
   <summary> ✔ Check the <b>services ip</b>, <b>vm</b>, and 2 <b>loadbalancer</b> status
   </summary>

   <img width="1918" height="1043" alt="Image" src="https://github.com/user-attachments/assets/55851997-7bb7-4213-b186-0abb812dfdd2" />
   <img width="1918" height="1043" alt="Image" src="https://github.com/user-attachments/assets/22ccbc1c-e12d-4888-bcff-6aa9b2d0bdd4" />
   <img width="1918" height="1043" alt="Image" src="https://github.com/user-attachments/assets/cbb04f01-efce-4285-974b-c07ec2e63386" />

   </details>

1. <details>
   <summary> ✔ Connect to the <b>loadbalancer</b>, the server can be connected.
   </summary>

   <img width="1911" height="1118" alt="Image" src="https://github.com/user-attachments/assets/c2af5731-5853-42fa-b420-445b04dcbd83" />
   <img width="1911" height="1118" alt="Image" src="https://github.com/user-attachments/assets/270b2663-5ac4-4d16-81a2-fb84a1db7fe0" />

   </details>
